src/core/atomic.d

```diff
@@ -1142,6 +1142,45 @@ version (CoreUnittest)
             assert(atomicOp!"+="(i64, 8) == 16);
         }
     }
+    
+    @betterC pure nothrow @nogc unittest
+    {
+        byte[] byteArray = [1, 3, 5, 7, 9, 11, 13, 15, 17, 19];
+        ulong[] ulongArray = [2, 4, 6, 8, 10, 12, 14, 16, 19, 20];
```

Comment:
`core.atomic` is meant to work with `shared` types. I think the functions that take non-`shared` arguments are for backwards compatibility with old code. Can you make these array variables `shared`, in order to be consistent with the general concept of this module?

```suggestion
        shared(ulong)[] ulongArray = [2, 4, 6, 8, 10, 12, 14, 16, 19, 20];
```

Reply:
I'm not sure. shared is still considered broken by some people. It also gets in the way of creating proper atomic variable like in [C++](https://en.cppreference.com/w/cpp/atomic).

Reply:
> I'm not sure. shared is still considered broken by some people.

The latest state of `shared` is that it's not broken, just incomplete. `core.atomic` and `core.sync.mutex` [1] work perfectly with `shared` as far as I know.

> It also gets in the way of creating proper atomic variable like in C++.

Do you have a bug report? As far as I can tell it works without issues based on a quick implementation I just wrote [2]

[1]: After my PR some time ago: https://github.com/dlang/druntime/pull/1728
[2]: https://gist.github.com/PetarKirov/8590707f5a21347dc4684b77d092cf60 /
https://run.dlang.io/gist/PetarKirov/8590707f5a21347dc4684b77d092cf60?compiler=dmd&args=-main%20-unittest%20-checkaction%3Dcontext / https://run.dlang.io/is/MsuNVh

Reply:
Interesting. I encountered these atomic issues when trying to implement my full version. It seems my issue stemmed from not having the shared attribute over the functions. I cannot find documentation stating that shared should be used as function attribute.
So this is a documentation issue. Does shared have any documentation other than [this](https://dlang.org/spec/attribute.html#shared)?

Reply:
> It seems my issue stemmed from not having the shared attribute over the functions.

Yes, analogous to C++ `const` member functions, in D a member function `func` must have the `shared` qualifier, if you want to call it on a `shared` object - `shared(Obj) obj; obj.func()`.

> I cannot find documentation stating that shared should be used as function attribute.

You can overload member functions on the type qualifier of `this`, whether it is `shared`, just like `const`, `immutable` and `inout`. After searching, the only mention of that language feature in the spec that I could find is this one: https://dlang.org/spec/class.html#member-functions

Reply:
> core.atomic is meant to work with shared types

No, `core.atomic` is supposed to _support_ shared types. Unshared has nothing to do with backwards compatibility, it's just normal testing.
I think tests should not have `shared` unless they are testing `shared` specifically. If tests wanted to be comprehensive, it could be implemented as a test matrix or something.

Reply:
So should I just remove the shared versions of the test?

Reply:
I wouldn't make it a primary test, because it's obscuring the thing that's being tested. But it doesn't hurt to make sure shared is also covered in tests because it's one of the combinatorial things that is often overlooked in testing, definitely should to be supported, and `shared` is tricky enough that it can not be assumed to work naturally without testing.

Reply:
OK. I removed the shared test in the recent commit. Seeing as testing both at the same time like earlier submission caused too much confusion. How do you suggest to add the shared test?